### PR TITLE
Ensure all funcs in types pkg are well guarded (Part 1)

### PR DESCRIPTION
**Pull request by melekes**
Refs #693

* [x] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [x] Wrote tests
* [x] Updated CHANGELOG.md

### CODE REVIEW

File: types/events.go [hunk 1]
```diff
@@ -10,22 +10,17 @@ import (
 
 // Reserved event types
 const (
-	EventBond              = "Bond"
```
Comment (ebuchman) [inline types/events.go]: omg lol! ancient events. That said we do want events for valset changes: https://github.com/tendermint/tendermint/issues/1916

File: types/genesis.go [hunk 2]
```diff
@@ -26,17 +26,7 @@ type GenesisDoc struct {
...
-}
-
-// AppState returns raw application state.
-// TODO: replace with AppState field during next breaking release (0.18)
-func (genDoc *GenesisDoc) AppState() json.RawMessage {
-	if len(genDoc.AppOptions) > 0 {
-		return genDoc.AppOptions
-	}
-	return genDoc.AppStateJSON
+	AppState        json.RawMessage    `json:"app_state,omitempty"`
```
Comment (ebuchman) [inline types/genesis.go]: Technically breaking on Go APIs but preserves the JSON. According to our latest SemVer, that's a non breaking version change. Cool.

File: types/block_test.go [hunk 3]
```diff
@@ -10,7 +10,25 @@ import (
...
+
+	voteSet, valSet, vals := randVoteSet(h-1, 1, VoteTypePrecommit, 10, 1)
+	commit, err := MakeCommit(lastID, h-1, 1, voteSet, vals)
+	require.NoError(t, err)
+
+	block := MakeBlock(h, txs, commit)
+	require.NotNil(t, block)
+
+	ev := NewMockGoodEvidence(h, 0, valSet.Validators[0].Address)
+	block.AddEvidence([]Evidence{ev})
```
Comment (ebuchman) [inline types/block_test.go]: shouldn't we finish the test by checking that the evidence got added?

**Comment by ebuchman** [comment]
Before:

```
types$ go test -cover
PASS
coverage: 54.1% of statements
ok      github.com/tendermint/tendermint/types  0.942s
```

After:

```
types$ go test -cover
PASS
coverage: 72.5% of statements
ok      github.com/tendermint/tendermint/types  1.055s
```

File: types/block_test.go (hunk 3 above)
Comment (melekes) [reply types/block_test.go]: yes

File: types/events.go (hunk 1 above)
Comment (melekes) [reply types/events.go]: we'll reintroduce them in #1916 then

**Pull request merged**